File: packages/java/parser-jvm-plugin-backbone/src/test/java/com/vaadin/hilla/parser/plugins/backbone/iterable/IterableEndpoint.java
package com.vaadin.hilla.parser.plugins.backbone.iterable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

@Endpoint
public class IterableEndpoint {
    public AnotherCustomIterable<Foo> getFooAnotherCustomIterable() {
        return new AnotherCustomIterable<>();
    }

    public Foo[] getFooArray() {
        return new Foo[] {};
    }

    public CustomIterable<Foo> getFooCustomIterable() {
        return new CustomIterable<>();
    }

    public Iterable<Foo> getFooIterable() {
        return Arrays.asList(new Foo(), new Foo());
    }

    public SpecializedIterable getSpecializedIterable() {
        return new SpecializedIterable();
    }

    public SpecializedIterableCustom getSpecializedIterableCustom() {
        return new SpecializedIterableCustom();
    }

    public List<Foo> getFooList() {
        return new ArrayList<>();
    }

    public Set<Foo> getFooSet() {
        return new HashSet<>();
    }

    // use a dedicated type `Bar` instead of reusing `Foo` to make sure it's
    // added to the OpenAPI schema
    public IterableWithProperties<Bar> getIterableWithProperties() {
        return new IterableWithProperties<>();
    }

    public static class IterableWithProperties<T> implements Iterable<T> {
        private final List<T> list = new ArrayList<>();
        private T defaultItem;

        public int size() {
            return list.size();
        }

        public T getDefaultItem() {
            return defaultItem;
        }

        public void setDefaultItem(T defaultItem) {
            this.defaultItem = defaultItem;
        }

        @Override
        public Iterator<T> iterator() {
            return list.iterator();
        }
    }

    public static class AnotherCustomIterable<T> implements Iterable<T> {
        private final List<T> list = new ArrayList<>();

        @Override
        public Iterator<T> iterator() {
            return list.iterator();
        }
    }

    public static class CustomIterable<T> implements Iterable<T> {
        private final List<T> list = new ArrayList<>();

        @Override
        public Iterator<T> iterator() {
            return list.iterator();
        }
    }

    public static class SpecializedIterable extends ArrayList<String> {
    }

    public static class SpecializedIterableCustom extends ArrayList<Foo> {
    }

    public static class Foo {
        public String bar = "bar";
    }

    // Used in `getIterableWithProperties` to make sure it's added to the
    // OpenAPI schema. Don't use it elsewhere.
    public static class Bar {
        public String foo = "foo";
    }
}

File: packages/java/typescript-generator/src/test/resources/com/vaadin/hilla/parser/plugins/backbone/iterable/snapshots/IterableEndpoint.ts
import type { EndpointRequestInit } from '@vaadin/hilla-frontend';
import type Foo from './com/vaadin/hilla/parser/plugins/backbone/iterable/IterableEndpoint/Foo.js';
import client from './connect-client.default.js';

export async function getFooAnotherCustomIterable(
  init?: EndpointRequestInit,
): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooAnotherCustomIterable', {}, init);
}

export async function getFooArray(init?: EndpointRequestInit): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooArray', {}, init);
}

export async function getFooCustomIterable(init?: EndpointRequestInit): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooCustomIterable', {}, init);
}

export async function getFooIterable(init?: EndpointRequestInit): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooIterable', {}, init);
}

export async function getFooList(init?: EndpointRequestInit): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooList', {}, init);
}

export async function getFooSet(init?: EndpointRequestInit): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getFooSet', {}, init);
}

export async function getSpecializedIterable(
  init?: EndpointRequestInit,
): Promise<Array<string | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getSpecializedIterable', {}, init);
}

export async function getSpecializedIterableCustom(
  init?: EndpointRequestInit,
): Promise<Array<Foo | undefined> | undefined> {
  return client.call('IterableEndpoint', 'getSpecializedIterableCustom', {}, init);
}
